packages/entity-server/src/routes/hierarchy/format/formatEntityForResponse.ts

```diff
@@ -0,0 +1,123 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType } from '../../../models';
+import { EntityServerModelRegistry } from '../../../types';
+import {
+  HierarchyContext,
+  ExtendedEntityFields,
+  FlattableEntityFields,
+  EntityResponseObject,
+} from '../types';
+import { extendedFieldFunctions, isExtendedField } from '../extendedFieldFunctions';
+import { EntityResponseObjectBuilder } from './EntityResponseObjectBuilder';
+
+export async function formatEntityForResponse(
+  ctx: HierarchyContext,
+  entity: EntityType,
+  fields: (keyof ExtendedEntityFields)[],
+): Promise<EntityResponseObject>;
+export async function formatEntityForResponse(
+  ctx: HierarchyContext,
+  entity: EntityType,
+  flat: keyof FlattableEntityFields,
+): Promise<EntityType[typeof flat]>;
+export async function formatEntityForResponse(
+  ctx: HierarchyContext,
+  entity: EntityType,
+  fieldsOrFlat: (keyof ExtendedEntityFields)[] | keyof FlattableEntityFields,
+) {
+  if (!Array.isArray(fieldsOrFlat)) {
+    const flat = fieldsOrFlat;
+    return entity[flat];
+  }
+  const fields = fieldsOrFlat;
+  const responseBuilder = new EntityResponseObjectBuilder();
+  for (let i = 0; i < fields.length; i++) {
+    const field = fields[i];
+    if (isExtendedField(field)) {
+      responseBuilder.set(field, await extendedFieldFunctions[field](entity, ctx));
+    } else {
+      responseBuilder.set(field, entity[field]);
+    }
+  }
+  return responseBuilder.build();
+}
+
+export async function formatEntitiesForResponse(
+  models: EntityServerModelRegistry,
+  ctx: HierarchyContext,
+  entities: EntityType[],
+  fields: (keyof ExtendedEntityFields)[],
+): Promise<EntityResponseObject[]>;
+export async function formatEntitiesForResponse(
+  models: EntityServerModelRegistry,
+  ctx: HierarchyContext,
+  entities: EntityType[],
+  flat: keyof FlattableEntityFields,
+): Promise<EntityType[typeof flat][]>;
```

Comment:
I find this function overloading syntax so ugly, yet also quite useful... hope its not too verbose...